gnd/src/main/java/com/google/android/gnd/repository/ProjectRepository.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.gnd.repository
+
+import com.google.android.gnd.model.Project
+import com.google.android.gnd.model.Role
+import com.google.android.gnd.model.User
+import com.google.android.gnd.model.feature.FeatureType
+import com.google.android.gnd.model.layer.Layer
+import com.google.android.gnd.persistence.local.LocalDataStore
+import com.google.android.gnd.persistence.local.LocalValueStore
+import com.google.android.gnd.persistence.remote.NotFoundException
+import com.google.android.gnd.persistence.remote.RemoteDataStore
+import com.google.android.gnd.rx.Loadable
+import com.google.android.gnd.rx.annotations.Cold
+import com.google.android.gnd.rx.annotations.Hot
+import com.google.android.gnd.ui.map.CameraPosition
+import com.google.android.gnd.util.toImmutableList
+import com.google.common.collect.ImmutableList
+import com.google.common.collect.ImmutableMap
+import io.reactivex.Flowable
+import io.reactivex.Single
+import io.reactivex.processors.BehaviorProcessor
+import io.reactivex.processors.FlowableProcessor
+import io.reactivex.processors.PublishProcessor
+import java8.util.Optional
+import timber.log.Timber
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Singleton
+
+private const val LOAD_REMOTE_PROJECT_TIMEOUT_SECS: Long = 15
+private const val LOAD_REMOTE_PROJECT_SUMMARIES_TIMEOUT_SECS: Long = 30
+
+/**
+ * Coordinates persistence and retrieval of [Project] instances from remote, local, and in
+ * memory data stores. For more details on this pattern and overall architecture, see
+ * https://developer.android.com/jetpack/docs/guide.
+ */
+@Singleton
+class ProjectRepository @Inject constructor(
+    private val userRepository: UserRepository,
+    private val localDataStore: LocalDataStore,
+    private val remoteDataStore: RemoteDataStore,
+    private val localValueStore: LocalValueStore) {
+
+    /** Emits a project id on {@see #activateProject} and empty on {@see #clearActiveProject}.  */
+    private val selectProjectEvent: @Hot FlowableProcessor<Optional<String>> = PublishProcessor.create()
```

Comment:
I thought `null` can't be passed to observables. That was the reason I kept it as `Optional`.

Reply:
That's correct, in fact all relevant methods are declared `@NonNull` in RxJava3:
http://reactivex.io/RxJava/3.x/javadoc/io/reactivex/rxjava3/core/Flowable.html

Reply:
ah that makes sense, pls ignore me then, lol, thanks!

Reply:
But it may also be clearer to keep it for now since Java callers don't have the same type system. Up to you.

Reply:
There is one use-case where we intentionally set `Optional.empty()` when clearing active project. So changing this seems non-trivial as of now.

Reply:
sounds good!

Reply:
It may not apply here, in general we should minimize the number of equivalent "empty" states; if we don't need to differentiate `optional.empty()` and `option.get().isEmpty()`, then getting rid of the outer `Optional` and relying on `String.isEmpty()` would simplify usage a potential for unhandled edge cases.

Reply:
That makes sense! We can replace `Optional.empty()` with an empty string "". Will update the changes.